src/transformers/models/whisper/tokenization_whisper.py

```diff
@@ -705,3 +705,278 @@ def get_decoder_prompt_ids(self, task=None, language=None, no_timestamps=True):
+            if "stride" in output:
+                chunk_len, stride_left, stride_right = output["stride"]
+
+                time_offset -= stride_left
+                right_stride_start = time_offset + (chunk_len - stride_right)
+
+            current_tokens = []
+
+            for i, token in enumerate(token_ids):
+                # 4 possible states for each token
+                # - 1/ Language code
+                # - 2/ all other special tokens (which we ignore)
+                # - 3/ Timestamp
+                # - 4/ Regular text
+                if token in self.all_special_ids:
+                    # Either language code or other
+                    text = self.decode([token])
+                    # Removing outer shell <|XX|>
+                    text = text[2:-2]
+                    language = LANGUAGES.get(text, None)
+                    if language is not None:
+                        # 1/ Indeed some language
+                        # TODO Handle when language is different from the previous
+                        # one, and we cannot use timestamped tokens to create chunks
+                        last_language = language
+                        chunk["language"] = language
+                    else:
+                        # 2/ This is a regular special token, ignoring it
+                        pass
+                elif token >= timestamp_begin:
+                    # 3/ Timestamp token
+                    time = (token - timestamp_begin) * time_precision + time_offset
+                    time = round(time, 2)
+                    if right_stride_start and time > right_stride_start:
+                        # Whisper outputted a timestamp token, but it falls within
+                        # our stride, so we're going to skip it for the time being
+                        # and resolve this later
+                        # Skip is necessary because timestamp tokens always come
+                        # by pair, so we need to skip the next one too.
+                        skip = True
+                        continue
+                    if chunk["timestamp"][0] is None:
+                        chunk["timestamp"][0] = time
+                    elif skip:
+                        skip = False
+                    else:
+                        # This is the end of the timestamp chunk
+                        chunk["timestamp"][1] = time
+                        previous_tokens.append(current_tokens)
+                        # XXX: Handling merges. We always give priority to the tokens
+                        # from the LAST chunk (whisper seems to perform better this way)
+                        # There is only a merge to handle when there is more than one
+                        # sequence here. That happens only when the last timestamp
+                        # happened in the stride of the audio
+                        resolved_tokens = _find_longest_common_sequence(previous_tokens)
```

Comment:
My only issue is that here we are looking at all the previous tokens, this is prone to a lot of error + slow

Reply:
This seems actually somewhat faster than before (not sure because `num_workers=1` in the bench seems to gain so much speedup).

This only takes the last *unused* tokens, so it won't take the entire previous audio for processing.

This function is indeed quite slow, but it's quite necessary to handle all the weird cases of timestamps outputs within whisper (like I've seen pretty much all cases now, no overlap, regular  ([a, b, c], [b, c, d]) left included in right ([b, c], [a, b, c]) right included in left ([a, b, c], [b, c]).

The parts where output is the worst is actually linked to the model where we get something like

([a, b, c, d], [b,c]]) and `d` is actually correct. But we cannot take it in general, because in general either `d` will be contained in the next token, or it is hallucinated and actually incorrect (but there are *some* cases where it's correct and the follow-up misses it)